lxd/identity/cache.go

```diff
@@ -0,0 +1,151 @@
+package identity
+
+import (
+	"crypto/x509"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/canonical/lxd/shared"
+	"github.com/canonical/lxd/shared/api"
+)
+
+// Cache represents an thread-safe in-memory cache of the identities in the database.
+type Cache struct {
+	entries map[string]map[string]CacheEntry
+	mu      sync.RWMutex
+}
+
+// CacheEntry represents an identity. The Certificate field is pre-computed for identities with AuthenticationMethod
+// api.AuthenticationMethodTLS.
+type CacheEntry struct {
+	Identifier           string
+	AuthenticationMethod string
+	IdentityType         string
+	Projects             []string
+	Certificate          *x509.Certificate
+}
+
+// Get returns a single CacheEntry by its authentication method and identifier.
+func (c *Cache) Get(authenticationMethod string, identifier string) (*CacheEntry, error) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	if c.entries == nil {
+		c.entries = make(map[string]map[string]CacheEntry)
```

Comment:
Initialising the map on a read (Get) seems a little odd. Can we just return not found in this case, and save the map initialisation for for the writes?

Reply:
Yeah this is weird and also incorrect since the mutex only has a read lock. Will fix.